elytron/src/main/java/org/wildfly/extension/elytron/FailoverRealmDefinition.java

```diff
@@ -0,0 +1,169 @@
+import org.jboss.as.controller.registry.AttributeAccess;
+import org.jboss.as.controller.registry.ManagementResourceRegistration;
+import org.jboss.as.controller.registry.OperationEntry;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.ModelType;
+import org.jboss.msc.inject.Injector;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController.Mode;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceTarget;
+import org.jboss.msc.value.InjectedValue;
+
+import org.wildfly.security.auth.realm.FailoverModifiableSecurityRealm;
+import org.wildfly.security.auth.realm.FailoverSecurityRealm;
+import org.wildfly.security.auth.server.ModifiableSecurityRealm;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityRealm;
+import org.wildfly.security.auth.server.event.SecurityRealmUnavailableEvent;
+
+import java.util.function.Consumer;
+
+/**
+ * A {@link ResourceDefinition} for a {@link SecurityRealm} which wraps one realm and fails over to another in case the first is unavailable.
+ *
+ * @author <a href="mailto:mmazanek@jboss.com">Martin Mazanek</a>
+ */
+class FailoverRealmDefinition extends SimpleResourceDefinition {
+
+    static final ServiceUtil<SecurityRealm> REALM_SERVICE_UTIL = ServiceUtil.newInstance(SECURITY_REALM_RUNTIME_CAPABILITY, ElytronDescriptionConstants.AGGREGATE_REALM, SecurityRealm.class);
+
+    static final SimpleAttributeDefinition DELEGATE_REALM = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.DELEGATE_REALM, ModelType.STRING, false)
+            .setMinSize(1)
+            .setCapabilityReference(SECURITY_REALM_CAPABILITY, SECURITY_REALM_CAPABILITY)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition FAILOVER_REALM = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.FAILOVER_REALM, ModelType.STRING, false)
+            .setMinSize(1)
+            .setCapabilityReference(SECURITY_REALM_CAPABILITY, SECURITY_REALM_CAPABILITY)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition EMIT_EVENTS = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.EMIT_EVENTS, ModelType.BOOLEAN, true)
+            .setAllowExpression(true)
+            .setMinSize(1)
+            .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+            .setDefaultValue(ModelNode.TRUE)
+            .build();
+
+    static final AttributeDefinition[] ATTRIBUTES = new AttributeDefinition[] { DELEGATE_REALM, FAILOVER_REALM, EMIT_EVENTS};
+
+    private static final AbstractAddStepHandler ADD = new RealmAddHandler();
+    private static final OperationStepHandler REMOVE = new TrivialCapabilityServiceRemoveHandler(ADD, SECURITY_REALM_RUNTIME_CAPABILITY);
+
+    FailoverRealmDefinition() {
```

Comment:
The realm should offer operations with identity (`add-identity`, `read-identity`, ...).

Reply:
I would only expect a realm to expose those operations if it exposes the modifiable realm capability, it wouldn't make sense to be able to support failing over things like an add-identity operation.